app/cdap/components/DirectiveInput/Components/AutoComplete/index.tsx

```diff
@@ -0,0 +1,236 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import { Box, Typography } from '@material-ui/core';
+import MyDataPrepApi from 'api/dataprep';
+import {
+  IAutoCompleteProps,
+  IOnRowClickValue,
+} from 'components/DirectiveInput/Components/AutoComplete/types';
+import { defaultFuseOptions } from 'components/DirectiveInput/constants';
+import { IUsageDirectives } from 'components/DirectiveInput/types';
+import ee from 'event-emitter';
+import Fuse from 'fuse.js';
+import reverse from 'lodash/reverse';
+import Mousetrap from 'mousetrap';
+import React, { useEffect, useState } from 'react';
+import globalEvents from 'services/global-events';
+import NamespaceStore from 'services/NamespaceStore';
+import styled from 'styled-components';
+import uuidV4 from 'uuid/v4';
+
+const DirectiveLabel = styled(Typography)`
+  font-style: normal;
+  font-weight: 600;
+  font-size: ${(props) => (props.isDescription ? '14px' : '16px')};
+  letter-spacing: 0.15;
+  color: #616161;
+`;
+
+const DirectiveResultRow = styled(Box)`
+  padding: 10px;
+  border-bottom: 1px solid #e0e0e0;
+  background-color: ${(props) => (props.isActive ? '#EFF0F2' : '#FFFFFF')};
+  &:hover {
+    background: #eff0f2;
+    cursor: pointer;
+  }
+`;
+
+export default function({
+  setDirectivesList,
+  isDirectiveSelected,
+  columnNamesList,
+  onRowClick,
+  getDirectiveUsage,
+  onColumnSelected,
+  directiveInput,
+}: IAutoCompleteProps) {
+  const [activeResults, setActiveResults] = useState<IUsageDirectives[]>([]);
+  const [input, setInput] = useState<string>('');
+  const [activeSelectionIndex, setActiveSelectionIndex] = useState<number | null>(null);
+  const eventEmitter = ee(ee);
+  const [fuse, setFuse] = useState(null);
+
+  const getUsage = () => {
+    if (!isDirectiveSelected) {
+      MyDataPrepApi.getUsage({ context: NamespaceStore.getState().selectedNamespace }).subscribe(
+        (res) => {
+          setDirectivesList(res.values);
+          setFuse(new Fuse(res.values, { ...defaultFuseOptions }));
+        }
+      );
+    } else {
+      setFuse(new Fuse(columnNamesList, { ...defaultFuseOptions, keys: ['label'] }));
+    }
+  };
+
+  useEffect(() => {
+    getUsage();
+  }, [isDirectiveSelected]);
+
+  useEffect(() => {
+    eventEmitter.on(globalEvents.DIRECTIVEUPLOAD, getUsage);
+    const directiveInput = document.getElementById('directive-input-search');
+    const mousetrap = new Mousetrap(directiveInput);
+    mousetrap.bind('up', handleUpArrow);
+    mousetrap.bind('down', handleDownArrow);
+    mousetrap.bind('enter', handleEnterKey);
+    mousetrap.bind('tab', handleTabKey);
+
+    // returned function will be called on component unmount
```

Comment:
this is known, what is happeing on the unmount
why are we using mousetrap and ewventemitter
plz explain properly

Reply:
When we start typing in input then a list appear based on input text, we navigate through that list using keyboard keys up and down, mousetrap and eventemitter are thus used for interaction of shortcut keys and events